### PR TITLE
[WIP] Swift-4 Upgrade

### PR DESCRIPTION
**Pull request by Vkt0r**
Hi guys,

Until XCode 9.0 beta 4 I was using the library without any problem using Carthage and compiling it using the XCode 9 command line tools when I ran `carthage update --platform ios`.

Now in XCode 9.0 beta 4 ocurred a minor change the API of `XCTestCase` the `name: String?` variable changed and the compiler was complaining about it when `carthage` ran. I was able to fix the compilation error removing the optional chaining.

After that I migrate the project to Swift 4 and set the default optimizations recommended by XCode, I fixed a minor issue regarding the migration and the library is compiling without any problem and it was possible to integrate in the project I'm using with XCode 9 beta4.

I think we can start migrating the library to Swift 4 as we can use libraries with other Swift versions without any problem thanks to XCode 9. I set the PR to `WIP` to work on any change necessary.

### CODE REVIEW

**Comment by Vkt0r** [comment]
Of course this not implies any change in `Bootstrap` that is where the test are failing

**Comment by ashfurrow** [comment]
Looks great so far! Let me know what help I can provide. I'm travelling over the next few weeks so if you're blocked on feedback, just ping me!

**Comment by Vkt0r** [comment]
@ashfurrow Thanks!! I'll let you know, so far the migration of the `Bootstrap` project is the more tricky I think because is where the unit tests are running, If I'm stuck I'll ask for help 😅.

File: Nimble_Snapshots/CurrentTestCaseTracker.swift [hunk 1]
```diff
@@ -17,12 +17,8 @@ import XCTest
 
 extension XCTestCase {
     var sanitizedName: String? {
-        guard let fullName = self.name else {
-            return nil
-        }
-
         let characterSet = CharacterSet(charactersIn: "[]+-")
-        let name = fullName.components(separatedBy: characterSet).joined()
+        let name = self.name.components(separatedBy: characterSet).joined()
```
Comment (marcgovi) [inline Nimble_Snapshots/CurrentTestCaseTracker.swift]: In Xcode 9 beta 6 they revert it back. `name` is an optional again 🙃

Comment (Vkt0r) [inline Nimble_Snapshots/CurrentTestCaseTracker.swift]: Interesting, thanks @marcgovi for the comment. I've been a little busy lately, I'l l try to continue the migration as far I can and update it for XCode 9 beta 6

Comment (Vkt0r) [inline Nimble_Snapshots/CurrentTestCaseTracker.swift]: @marcgovi I updated the project to use XCode 9 beta 6 and `self.name` is still non-optional, 😥. Please let me know if you can point me where did you saw it?? Thanks

Comment (marcgovi) [inline Nimble_Snapshots/CurrentTestCaseTracker.swift]: Pretty sure is an optional, you can also see it in #133 that has been merge it today. With beta 6 the compiler has to be complaining. Maybe just clean derived data + project and try again.

Comment (Vkt0r) [inline Nimble_Snapshots/CurrentTestCaseTracker.swift]: @marcgovi Yes right, something is weird in my machine 😬 I've to  check

**Comment by ashfurrow** [comment]
I think #137 covers this – thanks a lot for the PR!

**Comment by Vkt0r** [comment]
@ashfurrow No problem, you're welcome glad #137 covers it.

**Comment by ashfurrow** [comment]
👍 I've released this as version 6.3.0, as always please let me know if you have any issues.

**Pull request closed without merging**